numpy/lib/stride_tricks.py

```diff
@@ -34,15 +49,80 @@ def as_strided(x, shape=None, strides=None, subok=False):
     # Make sure dtype is correct in case of custom dtype
     if array.dtype.kind == 'V':
         array.dtype = x.dtype
-    if type(x) is not type(array):
-        # if input was an ndarray subclass and subclasses were OK,
-        # then view the result as that subclass.
-        array = array.view(type=type(x))
-        # Since we have done something akin to a view from x, we should let
-        # the subclass finalize (if it has it implemented, i.e., is not None).
-        if array.__array_finalize__:
-            array.__array_finalize__(x)
-    return array
+    return _maybe_view_as_subclass(x, array)
+
+
+def _broadcast_to(array, shape, subok, readonly):
+    shape = tuple(shape) if np.iterable(shape) else (shape,)
+    array = np.array(array, copy=False, subok=subok)
+    if not shape and array.shape:
+        raise ValueError('cannot broadcast a non-scalar to a scalar array')
+    if any(size < 0 for size in shape):
+        raise ValueError('all elements of broadcast shape must be non-'
+                         'negative')
+    broadcast = np.nditer((array,), flags=['multi_index', 'zerosize_ok'],
+                          itershape=shape, order='C').itviews[0]
+    result = _maybe_view_as_subclass(array, broadcast)
+    # nditer.itviews can be a read-only array, so we need to ensure the result
+    # is writeable
+    result.flags.writeable = not readonly
```

Comment:
I think it might be safer to do:
```
if readonly:
     result.flags.writable = False
```
Can you add a test (if not done already) that a readonly array works? I imagine this might throw an error for broadcast_arrays if the array is already readonly, or if not throw an error it might unset an existing readonly flag (I am not sure about the logic, but I think we cannot change the flag if the parent is readonly).